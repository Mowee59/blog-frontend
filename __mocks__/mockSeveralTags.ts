// New file: __mocks__/mockSeveralTags.ts
import { Article, Article_NoRelations } from '@/interfaces/Article';
import { Tag } from '@/interfaces/Tag';

export const mockSeveralTags: Tag[] = [
  {
    id: 1,
    attributes: {
      name: "React",
      locale: "fr",
      createdAt: "2024-09-11T08:17:24.849Z",
      updatedAt: "2024-09-27T10:18:29.586Z",
      description: "Une librairie Javascript pour construire des interfaces utilisateur",
      cover: {
        data: {
          id: 3,
          attributes: {
            name: "react.webp",
            alternativeText: "Image de couverture catégorie React",
            caption: "React",
            width: 6000,
            height: 4000,
            formats: {
              thumbnail: {
                name: "thumbnail_react.webp",
                hash: "thumbnail_react_8111d1aa1b",
                ext: ".webp",
                mime: "image/webp",
                path: '',
                width: 234,
                height: 156,
                size: 3.86,
                url: "/uploads/thumbnail_react_8111d1aa1b.webp"
              },
              small: {
                name: "small_react.webp",
                hash: "small_react_8111d1aa1b",
                ext: ".webp",
                mime: "image/webp",
                path: '',
                width: 300,
                height: 200,
                size: 5.21,
                url: "/uploads/small_react_8111d1aa1b.webp"
              },
              medium: {
                name: "medium_react.webp",
                hash: "medium_react_8111d1aa1b",
                ext: ".webp",
                mime: "image/webp",
                path: '',
                width: 750,
                height: 500,
                size: 15.35,
                url: "/uploads/medium_react_8111d1aa1b.webp"
              },
              large: {
                name: "large_react.webp",
                hash: "large_react_8111d1aa1b",
                ext: ".webp",
                mime: "image/webp",
                path: '',
                width: 1000,
                height: 667,
                size: 21.93,
                url: "/uploads/large_react_8111d1aa1b.webp"
              }
            },
            hash: "react_8111d1aa1b",
            ext: ".webp",
            mime: "image/webp",
            size: 3633.27,
            url: "/uploads/react_8111d1aa1b.webp",
            previewUrl: '',
            provider: "local",
            createdAt: "2024-09-11T08:17:11.574Z",
            updatedAt: "2024-09-20T09:36:16.491Z"
          }
        }
      },
      articles: {
        data: [{} as Article, {} as Article, {} as Article]
      },
      localizations: {
        data: []
      }
    }
  },
  {
    id: 2,
    attributes: {
      name: "TypeScript",
      locale: "fr",
      createdAt: "2024-09-12T10:20:30.123Z",
      updatedAt: "2024-09-28T11:22:33.456Z",
      description: "Un sur-ensemble typé de JavaScript qui se compile en JavaScript pur",
      cover: {
        data: {
          id: 4,
          attributes: {
            name: "typescript.webp",
            alternativeText: "Image de couverture catégorie TypeScript",
            caption: "TypeScript",
            width: 6000,
            height: 4000,
            formats: {
              thumbnail: {
                name: "thumbnail_typescript.webp",
                hash: "thumbnail_typescript_9222e2bb2c",
                ext: ".webp",
                mime: "image/webp",
                path: '',
                width: 234,
                height: 156,
                size: 4.12,
                url: "/uploads/thumbnail_typescript_9222e2bb2c.webp"
              },
              small: {
                name: "small_typescript.webp",
                hash: "small_typescript_9222e2bb2c",
                ext: ".webp",
                mime: "image/webp",
                path: '',
                width: 300,
                height: 200,
                size: 5.67,
                url: "/uploads/small_typescript_9222e2bb2c.webp"
              },
              medium: {
                name: "medium_typescript.webp",
                hash: "medium_typescript_9222e2bb2c",
                ext: ".webp",
                mime: "image/webp",
                path: '',
                width: 750,
                height: 500,
                size: 16.89,
                url: "/uploads/medium_typescript_9222e2bb2c.webp"
              },
              large: {
                name: "large_typescript.webp",
                hash: "large_typescript_9222e2bb2c",
                ext: ".webp",
                mime: "image/webp",
                path: '',
                width: 1000,
                height: 667,
                size: 24.56,
                url: "/uploads/large_typescript_9222e2bb2c.webp"
              }
            },
            hash: "typescript_9222e2bb2c",
            ext: ".webp",
            mime: "image/webp",
            size: 3987.65,
            url: "/uploads/typescript_9222e2bb2c.webp",
            previewUrl: '',
            provider: "local",
            createdAt: "2024-09-12T10:20:15.789Z",
            updatedAt: "2024-09-21T12:34:56.789Z"
          }
        }
      },
      articles: {
        data: [{} as Article, {} as Article]
      },
      localizations: {
        data: []
      }
    }
  },
  {
    id: 3,
    attributes: {
      name: "Next.js",
      locale: "fr",
      createdAt: "2024-09-13T14:25:36.789Z",
      updatedAt: "2024-09-29T15:27:39.012Z",
      description: "Un framework React pour la production",
      cover: {
        data: {
          id: 5,
          attributes: {
            name: "nextjs.webp",
            alternativeText: "Image de couverture catégorie Next.js",
            caption: "Next.js",
            width: 6000,
            height: 4000,
            formats: {
              thumbnail: {
                name: "thumbnail_nextjs.webp",
                hash: "thumbnail_nextjs_3333f3cc3d",
                ext: ".webp",
                mime: "image/webp",
                path: '',
                width: 234,
                height: 156,
                size: 3.98,
                url: "/uploads/thumbnail_nextjs_3333f3cc3d.webp"
              },
              small: {
                name: "small_nextjs.webp",
                hash: "small_nextjs_3333f3cc3d",
                ext: ".webp",
                mime: "image/webp",
                path: '',
                width: 300,
                height: 200,
                size: 5.43,
                url: "/uploads/small_nextjs_3333f3cc3d.webp"
              },
              medium: {
                name: "medium_nextjs.webp",
                hash: "medium_nextjs_3333f3cc3d",
                ext: ".webp",
                mime: "image/webp",
                path: '',
                width: 750,
                height: 500,
                size: 15.87,
                url: "/uploads/medium_nextjs_3333f3cc3d.webp"
              },
              large: {
                name: "large_nextjs.webp",
                hash: "large_nextjs_3333f3cc3d",
                ext: ".webp",
                mime: "image/webp",
                path: '',
                width: 1000,
                height: 667,
                size: 22.76,
                url: "/uploads/large_nextjs_3333f3cc3d.webp"
              }
            },
            hash: "nextjs_3333f3cc3d",
            ext: ".webp",
            mime: "image/webp",
            size: 3845.92,
            url: "/uploads/nextjs_3333f3cc3d.webp",
            previewUrl: '',
            provider: "local",
            createdAt: "2024-09-13T14:25:30.123Z",
            updatedAt: "2024-09-22T16:38:59.012Z"
          }
        }
      },
      articles: {
        data: [{} as Article, {} as Article, {} as Article, {} as Article]
      },
      localizations: {
        data: []
      }
    }
  }
];