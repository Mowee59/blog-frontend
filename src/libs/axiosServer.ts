import { Article } from "@/interfaces/article";
import { BlogHomePage } from "@/interfaces/blog-home-page";
import { Payload } from "@/interfaces/Payload";
import { Tag } from "@/interfaces/tag";
import axios, { AxiosResponse } from "axios";

/**
 *
 * This file is used to define api request that will be used for server side requests
 *
 */

// Creating axios instance
const apiClient = axios.create({
  // Fetching API url from env variables
  baseURL: process.env.API_URL,
  headers: {
    "Content-Type": "application/json",
  },
});

// // Add Axios interceptor for adding the authorization token
// apiClient.interceptors.request.use(
//   (config) => {
//     // Access the token from the environment variable
//     const token = process.env.STRAPI_API_TOKEN;

//     if (token) {
//       // Add the token to the Authorization header if it exists
//       config.headers.Authorization = `Bearer ${token}`;
//     }

//     return config;
//   },
//   (error) => {
//     return Promise.reject(error);
//   },
// );

/**
 * Fetch all the articles
 *
 * @returns A promise that resolves as a payload of Articles
 */
export const fetchArticles = async (): Promise<Payload<Article[]>> => {
  // TODO: Add axios typed response
  const response: AxiosResponse<Payload<Article[]>> = await apiClient.get(
    "/articles?status=published",
  );
  return response.data;
};

/**
 * Fetch the home page data
 *
 * @returns A promise that resolves as a payload containing home page data
 */
export const fetchHomePageData = async (): Promise<Payload<BlogHomePage>> => {
  // TODO: Add axios typed response
  const response = await apiClient.get(
    "/blog-home-page?populate[featuredArticle][populate][0]=coverImage",
  );
  return response.data;
};

/**
 * Fetch all the tags
 *
 * @returns A promise that resolves as a payload of Tags
 */
export const fetchTags = async (): Promise<Payload<Tag[]>> => {
  const response = await apiClient.get("/tags?populate=*");
  return response.data;
};

/**
 * Fetch a tag by its name
 *
 * @param name the name of the tag
 * @returns A promise that resolves as a Payload Tags
 */
export const fetchTagByName = async (name: string): Promise<Payload<Tag[]>> => {
  const response = await apiClient.get(
    `/tags?filters[name][$eqi]=${name}&populate=*`,
  );
  return response.data;
};

/**
 * Fetch all the articles associated with a specific tag name
 *
 * @param tagName Name of the tag
 * @params Optional parameters to add to te query
 * @returns
 */
export const fetchArticlesByTagName = async (
  tagName: string,
  params?: string,
): Promise<Payload<Partial<Article>[]>> => {
  const response = await apiClient.get(
    `/articles?filters[tags][name][$in]=${tagName}&${params ?? ""}`,
  );
  return response.data;
};

/**
 * Fetch an article by slug
 *
 * @param slug the slug of the article to retrieve
 * @returns A promise that resolves as a Payload containing a unique article
 */
export const fetchArticleBySlug = async (
  slug: string,
): Promise<Payload<Article[]>> => {
  const response = await apiClient.get(
    `/articles?filters[slug][$eqi]=${slug}&populate=*`,
  );
  return response.data;
};
