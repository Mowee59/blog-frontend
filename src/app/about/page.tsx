import { fetchAboutPageData } from "@/libs/axiosServer";
import markdownToHtml from "@/libs/markdownToHtml";
import { Metadata } from 'next';

export async function generateMetadata(): Promise<Metadata> {
  const fetchedData = await fetchAboutPageData();
  const seo = fetchedData.data.attributes.seo;

  return {
    title: seo?.metaTitle || 'À propos | Aniss.dev Blog',
    description: seo?.metaDescription || 'En savoir plus sur Aniss et son blog',
    openGraph: {
      title: seo?.metaTitle,
      description: seo?.metaDescription,
      images: seo?.metaImage?.data?.attributes.url ? [seo.metaImage.data.attributes.url] : [],
    },
  };
}

const About = async () => {
  try {
    const aboutPageDataPayload = await fetchAboutPageData();
    
    // Check if the about page data exists
    if (!aboutPageDataPayload.data) {
      throw new Error('About page data not found');
    }

    const aboutData = aboutPageDataPayload.data;

    // Parse the markdown content to HTML
    const content = await markdownToHtml(aboutData.attributes.content || '');

    return (
      <main className="container flex flex-col gap-8 lg:max-w-screen-lg">
        <div className="lg:px-10 xl:px-20">
          <h1 className="mb-6 text-2xl font-medium leading-normal text-neutral-900 dark:text-neutral-300 sm:text-4xl sm:leading-10 lg:mb-8 lg:text-5xl xl:text-6xl">
            À propos
          </h1>
          <div 
            className="prose prose-lg dark:prose-invert "
            dangerouslySetInnerHTML={{ __html: content || '' }}
          />
        </div>
      </main>
    );
    
  } catch (error) {
    console.error("Error fetching about page data:", error);
    return (
      <main className="container flex flex-col gap-8 lg:max-w-screen-lg">
        <div className="lg:px-10 xl:px-20">
          <h1 className="mb-6 text-2xl font-medium leading-normal text-neutral-900 dark:text-neutral-300 sm:text-4xl sm:leading-10 lg:mb-8 lg:text-5xl xl:text-6xl">
            Error
          </h1>
          <p>An error occurred while fetching the about page data. Please try again later.</p>
        </div>
      </main>
    );
  }
 
};

export default About;
