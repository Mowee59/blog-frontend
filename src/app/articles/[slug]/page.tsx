/*
 * This file is part of the Blog Frontend project.
 * 
 * Copyright (C) 2024 Aniss.dev
 * 
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */


import { Metadata } from "next";
import { fetchArticleBySlug, fetchArticles } from "@/libs/axiosServer";
import markdownToHtml from "@/libs/markdownToHtml";
import React from "react";
import "@/../public/css/prism.css";
import { notFound } from "next/navigation";

/**
 * This builtin in next js function allows to generate params so we can render pages on the server at build time and cache them this will improve performances and seo
 *
 */
export async function generateStaticParams() {
  const articlesPayload = await fetchArticles();
  const articles = articlesPayload.data;

  return articles.map((article) => ({
    slug: article.attributes.slug,
  }));
}

// This function generates metadata for the article page
export async function generateMetadata({
  params,
}: {
  params: { slug: string };
}): Promise<Metadata> {
  // Fetch the article data using the slug from the URL parameters
  const articlePayload = await fetchArticleBySlug(params.slug);

  // Check if the article exists
  if (!articlePayload.data || articlePayload.data.length === 0) {
    // If no article is found, return a default metadata object
    return {
      title: "Article Not Found",
      description: "The requested article could not be found.",
    };
  }

  const article = articlePayload.data[0];
  const seo = article.attributes.seo;

  // If there's no SEO data, return basic metadata
  if (!seo) {
    return {
      title: article.attributes.title,
      description: article.attributes.description,
    };
  }

  // If SEO data exists, return a more comprehensive metadata object
  return {
    // Basic metadata
    title: seo.metaTitle,
    description: seo.metaDescription,
    keywords: seo.keywords,
    robots: seo.metaRobots,
    viewport: seo.metaViewport,
    canonical: seo.canonicalURL,

    // Open Graph metadata for social media sharing
    openGraph: {
      title: seo.metaTitle || article.attributes.title,
      description: seo.metaDescription || article.attributes.description,
      url: `${process.env.NEXT_PUBLIC_SITE_URL}/articles/${article.attributes.slug}`,
      siteName: "Aniss.dev",
      images: seo.metaImage
        ? [
            {
              url: seo.metaImage.data.attributes.url,
              width: seo.metaImage.data.attributes.width,
              height: seo.metaImage.data.attributes.height,
              alt:
                seo.metaImage.data.attributes.alternativeText ||
                article.attributes.title,
            },
          ]
        : [],
      locale: "fr_FR",
      type: "article",
      publishedTime: article.attributes.publishedAt,
      authors: "Aniss",
    },

    // Twitter-specific metadata
    twitter: {
      card: "summary_large_image",
      title: seo.metaTitle || article.attributes.title,
      description: seo.metaDescription || article.attributes.description,
      images: seo.metaTwitterImage
        ? [
            {
              url: `${process.env.STRAPI_URL}${seo.metaTwitterImage.data.attributes.url}`,
              alt:
                seo.metaTwitterImage.data.attributes.alternativeText ||
                article.attributes.title,
              width: seo.metaTwitterImage.data.attributes.width,
              height: seo.metaTwitterImage.data.attributes.height,
            },
          ]
        : [],
    },

    // Include structured data if it exists
    ...(seo.structuredData && {
      other: {
        "script:ld+json": JSON.stringify(seo.structuredData),
      },
    }),
  };
}




// Generate page on demand if path hasn't been regenerated yet
export const dynamicParams = true;

const Article = async ({ params }: { params: { slug: string } }) => {
  try {
    // Retrieving the article as a strapi payload
    const articlePayload = await fetchArticleBySlug(params.slug);

    // Check if the article exists
    if (!articlePayload.data || articlePayload.data.length === 0) {
      // If no article is found, redirect to the 404 page
      notFound();
    }

    // Assigning the article data itself to a variable
    const article = articlePayload.data[0];
    // Creating a new Date object from the publishedAt attribute
    const publicationDate = new Date(article.attributes.publishedAt as string);

    // The array of tags the article is related with
    const tags = article.attributes.tags?.data ?? [];

    // We parse the markdown to get HTML content
    const parsedMarkdown = await markdownToHtml(article.attributes.content);

    return (
      <main className="container flex flex-col lg:max-w-screen-lg ">
        <div className=" lg:px-10 xl:px-20">
          <div className="flex flex-col">
            <div className="flex flex-col gap-6">
              <div className="flex justify-between">
                <h6 className="text-xs font-medium leading-3 text-neutral-600 dark:text-neutral-400">
                  {`${publicationDate.toLocaleDateString("Fr-fr", {
                    day: "numeric",
                    month: "long",
                    year: "numeric",
                  })}`}
                  <br />
                  {tags.map((tag) => (
                    <span key={tag.id} className="mr-1 hover:underline ">
                      {"#" + tag.attributes.name + " "}
                    </span>
                  ))}
                </h6>
                <h6 className="text-xs font-medium leading-3 text-neutral-600 dark:text-neutral-400">
                  {article.attributes.readingTime.slice(0, -5)}
                </h6>
              </div>
              <h1 className="mb-6 text-2xl font-medium leading-normal text-neutral-900 dark:text-neutral-300 sm:text-4xl sm:leading-10 lg:mb-8 lg:text-5xl xl:text-6xl">
                {article.attributes.title}
              </h1>
              {/* <p className="text-base leading-relaxed text-neutral-700 dark:text-neutral-400">
                {article.attributes.description}
              </p> */}
            </div>
          </div>
          <article
            dangerouslySetInnerHTML={{ __html: parsedMarkdown }}
            className="prose prose-slate dark:prose-invert lg:prose-xl prose-img:rounded-xl"
          ></article>
        </div>
      </main>
    );
  } catch (error) {
    console.error("Error fetching article:", error);
    notFound();
  }
};

export default Article;
