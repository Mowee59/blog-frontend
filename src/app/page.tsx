import ArticleCard from "@/components/ui/article-card/ArticleCard";
import Image from "next/image";
import hero from "/public/svg/hero.jpg";
import { fetchHomePageData } from "@/libs/axiosServer";
import ArticleList from "@/components/ui/article-list/ArticleList";

const Home = async () => {
  //Gettting the home page data
  const fetchedData = await fetchHomePageData();

  // Assigning the featuresd article to a variable for better readability
  const featuredArticle = fetchedData.data.attributes.featuredArticle?.data;
  // TODO Make featuredARticle required
  // Assigning the featuresd article's image to a variable for better readability
  const featuredImage = featuredArticle?.attributes.coverImage.data;

  return (
    <main className="container flex flex-col gap-8 lg:max-w-screen-lg">
      <div className=" lg:px-10 xl:px-20">
        <h1 className="mb-6 text-2xl font-medium leading-normal text-neutral-900 dark:text-neutral-300 sm:text-4xl sm:leading-10 lg:mb-8 lg:text-5xl xl:text-6xl">
          {featuredArticle?.attributes.title}
        </h1>
        <p className="text-base font-medium text-neutral-700 dark:text-neutral-400 sm:text-lg sm:leading-7 lg:text-xl lg:leading-loose">
          {featuredArticle?.attributes.description}
        </p>
      </div>
      <div>
        <div className="relative mx-auto  mb-4 h-[340px] w-full lg:h-[500px] ">
          <Image
            // Image url structure is : strapiUrl/uplooads/image
            src={`${featuredImage?.attributes.url}`}
            alt={featuredImage?.attributes.alternativeText || "cover image"}
            sizes="100vw"
            fill
            className=" rounded-xl object-cover"
          />
        </div>
        <p className="text-center text-xs font-normal italic leading-none text-neutral-600 dark:text-neutral-500">
          {featuredImage?.attributes.caption}
        </p>
      </div>
      <div className=" lg:px-10 xl:px-20">
        <ArticleList />
      </div>
    </main>
  );
};

export default Home;
