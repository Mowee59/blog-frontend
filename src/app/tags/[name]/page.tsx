import { fetchTagByName, fetchTags } from "@/libs/axiosServer";
import React from "react";
import Image from "next/image";
import ArticleList from "@/components/ui/article-list/ArticleList";
import { Metadata } from "next/types";
import { notFound } from 'next/navigation';

/**
 * This builtin in next js function allows to generate params so we can render pages on the server at build time and cache them this will improve performances and seo
 *
 */
export async function generateStaticParams() {
  const tagsPayload = await fetchTags();
  const tags = tagsPayload.data;

  return tags.map((tag) => ({
    name: tag.attributes.name,
  }));
}


// This function generates metadata for the tag page
export async function generateMetadata({ params }: { params: { name: string } }): Promise<Metadata> {
  try {
    // Fetch the tag data using the name from the URL parameters
    const tagPayload = await fetchTagByName(params.name);

    // Check if the tag exists
    if (!tagPayload.data || tagPayload.data.length === 0) {
      // If no tag is found, return arbitrary metadata for not found
      return {
        title: 'Tag Not Found',
        description: 'The requested tag could not be found.',
      };
    }

    const tag = tagPayload.data[0];
    const seo = tag.attributes.seo;

    // If there's no SEO data, return basic metadata
    if (!seo) {
      return {
        title: `#${tag.attributes.name}`,
        description: tag.attributes.description,
      };
    }

    // If SEO data exists, return a more comprehensive metadata object
    return {
      // Basic metadata
      title: seo.metaTitle,
      description: seo.metaDescription,
      keywords: seo.keywords,
      robots: seo.metaRobots,
      viewport: seo.metaViewport,
      canonical: seo.canonicalURL,

      // Open Graph metadata for social media sharing
      openGraph: {
        images: seo.metaImage ? [seo.metaImage.data.attributes.url] : [],
      },

      // Twitter-specific metadata
      twitter: {
        card: 'summary_large_image',
        images: seo.metaTwitterImage ? [seo.metaTwitterImage.data.attributes.url] : [],
      },

      // Include structured data if it exists
      ...(seo.structuredData && { 
        other: {
          'script:ld+json': JSON.stringify(seo.structuredData),
        },
      }),
    };
  } catch (error) {
    console.error("Error generating metadata for tag:", error);
    // Return arbitrary metadata for error case
    return {
      title: 'Error Loading Tag',
      description: 'There was an error loading the requested tag.',
    };
  }
}


// TODO Handle 404
// Generate page on demand if path hasn't been regenerated yet
export const dynamicParams = true;

const TagPage = async ({ params }: { params: { name: string } }) => {
  try {
    // Retrieving the tag as a strapi payload
    const tagPayload = await fetchTagByName(params.name);

    // Check if the tag exists
    if (!tagPayload.data || tagPayload.data.length === 0) {
      // If no tag is found, redirect to the 404 page
      notFound();
    }

    // Assigning the tag data to a variable
    const tag = tagPayload.data[0];

    const coverImage = tag.attributes.cover.data;

    return (
      <main className="container flex flex-col gap-8 lg:max-w-screen-lg">
        <div className=" lg:px-10 xl:px-20">
          <h1 className="mb-6 text-2xl font-medium leading-normal text-neutral-900 dark:text-neutral-300 sm:text-4xl sm:leading-10 lg:mb-8 lg:text-5xl xl:text-6xl">
            {`#${tag.attributes.name}`}
          </h1>
          <p className="text-base font-medium text-neutral-700 dark:text-neutral-400 sm:text-lg sm:leading-7 lg:text-xl lg:leading-loose">
            {tag.attributes.description}
          </p>
        </div>
        {/* #TODO Handle image sizing */}
        <div>
          <div className="relative mx-auto  mb-4 h-[340px] w-full lg:h-[500px] ">
            <Image
              // Image url structure is : strapiUrl/uplooads/image
              src={`${process.env.STRAPI_URL}${coverImage.attributes.url}`}
              alt={`${coverImage.attributes.alternativeText}` || "cover image"}
              fill
              className=" rounded-xl object-cover"
            />
          </div>
          <p className="text-center text-xs font-normal italic leading-none text-neutral-600 dark:text-neutral-500">
            {coverImage.attributes.caption}
          </p>
        </div>
        <div className=" lg:px-10 xl:px-20">
          <ArticleList
            queryParams={`filters[tags][name][$in]=${tag.attributes.name}`}
          />
        </div>
      </main>
    );
  } catch (error) {
    console.error("Error fetching tag:", error);
    notFound();
  }
};

export default TagPage;
