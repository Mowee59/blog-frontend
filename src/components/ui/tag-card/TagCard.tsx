import { Tag } from "@/interfaces/tag";
import React from "react";
import Image from "next/image";
import { fetchArticlesByTagName } from "@/libs/api";

//Typing the component props
type TagCardProps = {
  tag: Tag;
};

const TagCard = async (props: TagCardProps) => {
  // Assigning the cover image attributes to a variable
  const coverImageAttributes = props.tag.attributes.cover.data.attributes;

  // Asssigning the thumbnail to a varialbe
  const coverImageThumbnail = coverImageAttributes.formats.thumbnail;

  // Fetching the articles related to the tag
  // We are adding  the 'fields[0]' query parameter to filter all list car we are only interested in metadata ( to get the number of articles related to the tag)
  const articles = await fetchArticlesByTagName(
    props.tag.attributes.name,
    "fields[0]",
  );

  return (
    <article className="flex h-[225px] w-full flex-col gap-5">
      <div className="relative h-[150px]">
        <Image
          src={`${process.env.STRAPI_URL}${coverImageThumbnail.url}`}
          alt={coverImageAttributes.alternativeText}
          fill
          placeholder="blur"
          blurDataURL={`${process.env.STRAPI_URL}${coverImageAttributes.previewUrl}`}
          objectFit="cover"
          className=" rounded-xl"
        ></Image>
      </div>
      <div className="flex flex-col gap-1">
        <h3 className="text-xl font-medium leading-snug text-neutral-900 dark:text-neutral-300">{`#${props.tag.attributes.name}`}</h3>
        <h4 className="text-sm font-medium leading-4 text-[#3d3d3d] dark:text-neutral-400">{`${articles.meta.pagination?.total} posts`}</h4>
      </div>
    </article>
  );
};

export default TagCard;
