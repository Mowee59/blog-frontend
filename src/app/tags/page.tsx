import React from "react";
import { fetchArticles, fetchTags } from "@/libs/axiosServer";
import TagCard from "@/components/ui/tag-card/TagCard";
import ArticleList from "@/components/ui/article-list/ArticleList";

const Tags: React.FC = async () => {
  // Fetching all the tags
  const tagsPayload = await fetchTags();

  // we extract the actual tags list from the payload
  const tags = tagsPayload.data;

  return (
    <main className="container  lg:max-w-screen-lg">
      <section className="flex flex-col items-center gap-9 px-6 sm:grid sm:grid-cols-2 sm:gap-10 lg:grid-cols-3">
        {tags.map((tag) => (
          <TagCard tag={tag} key={tag.id}></TagCard>
        ))}
      </section>
      <section className="container  lg:max-w-screen-lg">
        <ArticleList />
      </section>
    </main>
  );
};

export default Tags;

//Invalidate the cache  every 600 secondes (10min)
export const revalidate = 600;
