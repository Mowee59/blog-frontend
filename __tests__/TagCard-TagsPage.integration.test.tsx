import React from "react";
import { render, screen } from "@testing-library/react";
import Tags from "../src/app/tags/page";
import { fetchTags } from "@/libs/axiosServer";
import { mockSeveralTags } from "../__mocks__/mockSeveralTags";
import TagCard from "@/components/ui/tag-card/TagCard";
import { generateMetadata } from "../src/app/tags/[name]/page";
import { fetchTagByName } from "@/libs/axiosServer";

// Mock the fetchTags function
jest.mock("@/libs/axiosServer", () => ({
  fetchTags: jest.fn(),
}));



// Mock the fetchTags function to return mockSeveralTags
(fetchTags as jest.Mock).mockResolvedValue({ data: mockSeveralTags });

/**
 * Integration tests for the Tags Page component
 */
describe("Tags Page and TagCard Integration Tests", () => {
  /**
   * Test case: Verify that all tags are rendered
   */
  it("renders all tags", async () => {
    // Arrange: Set up the expected tag count
    const expectedTagCount = mockSeveralTags.length;

    // Act: Render the Tags component and find all article elements
    const { findAllByRole } = render(await Tags({}));
    const tagCards = await findAllByRole("article");

    // Assert: Check if the number of rendered tag cards matches the expected count
    expect(tagCards).toHaveLength(expectedTagCount);
  });

  /**
   * Test case: Verify that TagCard components are rendered with correct data
   */
  it("renders TagCard components with correct data", async () => {
    // Arrange: Set up the expected tags data
    const expectedTags = mockSeveralTags;

    // Act: Render the Tags component
    render(await Tags({}));

    // Assert: Check if each tag's data is correctly rendered
    for (const tag of expectedTags) {
      // Check if tag name is rendered
      expect(screen.getByText(`#${tag.attributes.name}`)).toBeInTheDocument();
      // Check if article count is rendered
      expect(
        screen.getByText(`${tag.attributes.articles?.data.length} Articles`),
      ).toBeInTheDocument();

      // Check if tag image is rendered with correct attributes
      const image = screen.getByAltText(
        tag.attributes.cover.data.attributes.alternativeText,
      );
      expect(image).toBeInTheDocument();
      expect(image).toHaveAttribute(
        "src",
        expect.stringContaining(
          tag.attributes.cover.data.attributes.formats.thumbnail.url,
        ),
      );
    }
  });

  /**
   * Test case: Verify that TagCard components have correct links
   */
  it("renders TagCard components with correct links", async () => {
    // Arrange: Set up the expected tags data
    const expectedTags = mockSeveralTags;

    // Act: Render the Tags component and get all article elements
    render(await Tags({}));
    const tagCards = screen.getAllByRole("article");

    // Assert: Check if each tag card has the correct link
    for (const tag of expectedTags) {
      // Find the tag card for the current tag
      const tagCard = tagCards.find((card) =>
        card.textContent?.includes(`#${tag.attributes.name}`),
      );
      expect(tagCard).toBeTruthy();

      if (tagCard) {
        // Check if the link in the tag card has the correct href attribute
        const link = tagCard.querySelector("a");
        expect(link).toHaveAttribute(
          "href",
          `/tags/${tag.attributes.name.toLowerCase()}`,
        );
      }
    }
  });

  /**
   * Test case: Verify the correct number of TagCard components are rendered
   */
  it("renders the correct number of TagCard components", async () => {
    // Arrange: Set up the expected tag count
    const expectedTagCount = mockSeveralTags.length;

    // Act: Render the Tags component and get all article elements
    render(await Tags({}));
    const tagCards = screen.getAllByRole("article");

    // Assert: Check if the number of rendered tag cards matches the expected count
    expect(tagCards).toHaveLength(expectedTagCount);
  });

  /**
   * Test case: Verify that the Tags component matches the snapshot
   */
  it("matches snapshot", async () => {
    // Arrange & Act: Render the Tags component
    const { asFragment } = render(await Tags({}));

    // Assert: Check if the rendered component matches the snapshot
    expect(asFragment()).toMatchSnapshot();
  });

 
});
