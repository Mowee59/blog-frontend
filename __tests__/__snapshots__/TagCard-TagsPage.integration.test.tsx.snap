// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tags Page and TagCard Integration Tests matches snapshot 1`] = `
<DocumentFragment>
  <main
    class="container  min-h-screen lg:max-w-screen-lg"
  >
    <section
      class="flex flex-col items-center gap-9 px-6 sm:grid sm:grid-cols-2 sm:gap-10 lg:grid-cols-3"
    >
      <article
        class="flex h-[225px] w-full flex-col gap-5"
      >
        <a
          href="/tags/react"
        >
          <div
            class="flex flex-col gap-5"
          >
            <div
              class="relative h-[150px]"
            >
              <img
                alt="Image de couverture catégorie React"
                class=" rounded-xl object-cover"
                src="undefined/uploads/thumbnail_react_8111d1aa1b.webp"
              />
            </div>
            <div
              class="flex flex-col gap-1"
            >
              <h3
                class="text-xl font-medium leading-snug text-neutral-900 dark:text-neutral-300"
              >
                #React
              </h3>
              <h4
                class="text-sm font-medium leading-4 text-[#3d3d3d] dark:text-neutral-400"
              >
                3 Articles
              </h4>
            </div>
          </div>
        </a>
      </article>
      <article
        class="flex h-[225px] w-full flex-col gap-5"
      >
        <a
          href="/tags/typescript"
        >
          <div
            class="flex flex-col gap-5"
          >
            <div
              class="relative h-[150px]"
            >
              <img
                alt="Image de couverture catégorie TypeScript"
                class=" rounded-xl object-cover"
                src="undefined/uploads/thumbnail_typescript_9222e2bb2c.webp"
              />
            </div>
            <div
              class="flex flex-col gap-1"
            >
              <h3
                class="text-xl font-medium leading-snug text-neutral-900 dark:text-neutral-300"
              >
                #TypeScript
              </h3>
              <h4
                class="text-sm font-medium leading-4 text-[#3d3d3d] dark:text-neutral-400"
              >
                2 Articles
              </h4>
            </div>
          </div>
        </a>
      </article>
      <article
        class="flex h-[225px] w-full flex-col gap-5"
      >
        <a
          href="/tags/next.js"
        >
          <div
            class="flex flex-col gap-5"
          >
            <div
              class="relative h-[150px]"
            >
              <img
                alt="Image de couverture catégorie Next.js"
                class=" rounded-xl object-cover"
                src="undefined/uploads/thumbnail_nextjs_3333f3cc3d.webp"
              />
            </div>
            <div
              class="flex flex-col gap-1"
            >
              <h3
                class="text-xl font-medium leading-snug text-neutral-900 dark:text-neutral-300"
              >
                #Next.js
              </h3>
              <h4
                class="text-sm font-medium leading-4 text-[#3d3d3d] dark:text-neutral-400"
              >
                4 Articles
              </h4>
            </div>
          </div>
        </a>
      </article>
    </section>
  </main>
</DocumentFragment>
`;
